pallets/circuit/src/lib.rs

```diff
@@ -372,53 +374,38 @@ pub mod pallet {
-                        Some(xtx_id),
-                    )
-                        .unwrap();
-
-                    // Ensure Circuit::PendingBidding status
-                    if local_ctx.xtx.status != CircuitStatus::PendingBidding {
-                        Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
-                        return
-                    }
-
-                    let current_step = local_ctx.xtx.steps_cnt.0;
-                    for mut fsx in local_ctx.full_side_effects[current_step as usize].iter_mut() {
-                        let sfx_id = fsx.generate_id::<SystemHashing<T>, T>(local_ctx.xtx_id);
-                        if let Some(best_sfx_bid) = <PendingSFXBids<T>>::get(xtx_id, sfx_id) {
-                            fsx.best_bid = Some(best_sfx_bid);
-                        } else {
-                            // error - some FSX don't have bids
-                            Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
+                    Machine::<T>::compile_infallible(
+                        &mut Machine::<T>::load_xtx(xtx_id).expect("xtx_id corresponds to a valid Xtx when reading from PendingXtxBidsTimeoutsMap storage"),
+                        |current_fsx, _local_state, _steps_cnt, status, _requester| {
+                            if status != CircuitStatus::PendingBidding {
+                                return PrecompileResult::TryKill(Cause::Timeout)
+                            }
+                            for mut fsx in current_fsx.iter_mut() {
+                                let sfx_id = fsx.generate_id::<SystemHashing<T>, T>(xtx_id);
+                                // Either assign best bid to FSX or Kill entire Xtx with DroppedAtBidding cause.
+                                if let Some(best_sfx_bid) = <PendingSFXBids<T>>::get(xtx_id, sfx_id)
+                                {
+                                    fsx.best_bid = Some(best_sfx_bid);
+                                } else {
+                                    // Error - at least one FSX has no bid
+                                    return PrecompileResult::TryKill(Cause::Timeout)
+                                }
+                            }
+                            PrecompileResult::UpdateFSX(current_fsx.clone())
+                        },
+                        |status_change, local_ctx| {
+                            // Account fees and charges
+                            Self::square_up(local_ctx, status_change, None).expect(
+                                "Expect square up at DroppedAtBidding loop to be infallible",
+                            );
                             Self::emit_status_update(
                                 local_ctx.xtx_id,
-                                Some(local_ctx.xtx),
+                                Some(local_ctx.xtx.clone()),
                                 None,
                             );
```

Comment:
I think we need to move the `can_withdraw` step into `compile`. This way, then and only then, when all validation steps pass, can we apply the function to commit.

So compile or something before should ensure that ANYTHING that can fail can pass, then we apply F.

To do this, I suggest:
- remove result return time from square up
- `PSFX::get()` -> ensure it exists first in compile and just unwrap/expect
- check status is not equal to pendingbidding first

You can provide some functor which would do this validation first, if you don't want to hide it in `compile`.

So two HOF's:
- one to validate
- then to commit the changes after compile

Reply:
we can split the current square up operations to fallible and infallible:

- A) fallible lock requester for all SFX max rewards @CircuitStatus::Requested
- B) fallible lock executor at bidding for SFX @CircuitStatus::PendingBidding | @CircuitStatus::InBidding
- C) fallible executor's attempt to execute SFX via XBI @CircuitStatus::PendingBidding

- D)  (expected to be infallible) unreserve requester's max rewards and reserve requester's rewards for bid amounts @CircuitStatus::Ready
- E) infallible unreserve requester's max rewards and executor's bid amounts @CircuitStatus::Killed(_)
- F) infallible unreserve requester's max rewards and slash dishonest executors @CircuitStatus::Revert(_)
- G) infallible rewards payouts via AccountManager::finalize and infallible unlock executor's bonds @CircuitStatus::Finalize

I could move infallible D-G as part of compile / apply as they're all infallible unreserve / finalize 

then separate fallible A-C into another module and use separate by fn on_extrinsics_trigger, fn execute_sfx_via_xbi, fn bid_sfx in pre-compilation validation calls